Switch default implementation to v2 (#402)
GoogleCloudPlatform/cloud-spanner-r2dbc: README.md

# Cloud Spanner R2DBC Driver

This project contains:
* An implementation of Java Reactive Relational Database Connectivity SPI [R2DBC](https://r2dbc.io/) for [Cloud Spanner](https://cloud.google.com/spanner/) based on the Cloud Spanner [client library](https://github.com/googleapis/java-spanner).

> * A [Spring Data R2DBC dialect for Cloud Spanner](https://github.com/GoogleCloudPlatform/cloud-spanner-r2dbc/tree/main/cloud-spanner-spring-data-r2dbc).

Reviewer: It's a separate dependency, right?

How do we guide users to Spring Data R2DBC more? There doesn't seem to be much discussion on it in this doc.

Author: Right; I want to write a blurb about this also, as a follow-up PR. We have the sample, but it would be good to show in the docs.

The reason there was not any discussion of Spring Data before is that none of it is specific to Cloud Spanner driver.

Author: #407 to track.

```
dependencies {
  compile group: 'com.google.cloud', name: 'cloud-spanner-r2dbc', version: '0.5.0'
}
```

### Usage

After setting up the dependency and [authentication](#authentication), you can begin directly using the driver.

The rest of this documentation will show examples of directly using the driver.
In a real application, you should use one of R2DBC's user-friendly [client APIs](https://r2dbc.io/clients/) instead.

> The entry point to using the R2DBC driver is to first configure the R2DBC connection factory.

Reviewer 2: ```suggestion
The entrypoint to using the R2DBC driver is to first configure the R2DBC connection factory.
```

Is entrypoint one word?

Author: I'll just replace it by "To start".

Object-typed options can only be passed in programmatically.

|Property name            |Type                              |Allowed in URL connection |Default |Comments|
|-------------------------|-------|--------------------------|--------|--------|
|`credentials`            |String |Yes                       |null    |The location of the credentials file to use for this connection
|`oauthToken`             |String |Yes                       |null    |A valid pre-existing OAuth token to use for authentication
|`google_credentials`     |com.google.auth.oauth2.OAuth2Credentials|No|null|A pre-authenticated authentication object that can only be supplied with programmatic connection options
|`usePlainText`           |boolean|Yes                       |false   |Turns off SSL and credentials use (only valid when using Cloud Spanner emulator)
|`optimizerVersion`       |String |Yes                       |null    |Determines version of Cloud Spanner https://cloud.google.com/spanner/docs/query-optimizer/query-optimizer-versions[optimizer] to use in queries
|`autocommit`             |boolean|Yes                       |true    |Whether new connections are created in autocommit mode

> |`readonly`               |boolean|Yes                       |false   | Whether new connections start with a read-only transaction

Reviewer 2: ```suggestion
|`readonly`               |boolean|Yes                       |false   |Whether new connections start with a read-only transaction
```

    
    To customize staleness, pass in a `TimestampBound` parameter.
    See the [TransactionOptions documentation](https://cloud.google.com/spanner/docs/reference/rpc/google.spanner.v1#google.spanner.v1.TransactionOptions)
    for more information about all of the transaction type settings that are available.

    ```java
    Mono.from(connectionFactory.create())
                .flatMapMany(c ->
                    Flux.concat(
                              ((SpannerConnection) conn).beginReadonlyTransaction(TimestampBound.ofExactStaleness(1, TimeUnit.SECONDS)),
                                ...
                              conn.commitTransaction(),
                        )
    ```
    NOTE: Readonly transactions must be closed by calling `commit()` before starting a new read-write or a read-only transaction.

- Partitioned DML: A transaction designed for bulk updates and deletes with certain restrictions.
    See the [Partitioned DML documentation](https://cloud.google.com/spanner/docs/dml-partitioned)
    for more information.

>     This driver does not support Partitioned DML transactions at the time.

Reviewer: Is there an issue for this?

Author: I've created #406. Also forwarded a discussion from last year to the team.

Note that calling `execute` produces R2DBC `Result` objects, but this doesn't cause the query to be run on the database. 
You must use the `map` or `getRowsUpdated` methods of the results to complete the underlying queries.

## Backpressure

Backpressure on SQL SELECT queries is supported out of the box.

> Take care to ultimately always exhaust or cancel the query result `Publisher`, since not doing so may lead to objects not being deallocated properly.

Reviewer: ```suggestion
Take care to always ultimately exhaust or cancel the query result `Publisher`, since not doing so may lead to objects not being deallocated properly.
```

```

## Using Connection Pool

Client Spanner client library maintains its own low-level connection pool, making use of [r2dbc pool](https://github.com/r2dbc/r2dbc-pool) unnecessary.
When R2DBC connections are closed, the underlying Client Spanner connection is reused internally.

> ## gRPC-based driver (v1)

Reviewer: What's the rationale for still including v1? When is it going to be removed?

Author: We can just remove it as part of this change, I suppose. There is no instance in which we'd want customers using v1.

| `project`   | Your GCP Project ID        | True (if `url` not provided) | |
| `instance`  | Your Spanner Instance name | True (if `url` not provided) | |
| `database`  | Your Spanner Database name | True (if `url` not provided) | |
| `url`       | A Cloud Spanner R2DBC URL specifying your Spanner database. An alternative to specifying `project`, `instance`, and `database` separately. | False |
| `google_credentials` | Optional [Google credentials](https://cloud.google.com/docs/authentication/production) override to specify for your Google Cloud account. | False | If not provided, credentials will be [inferred from your runtime environment](https://cloud.google.com/docs/authentication/production#finding_credentials_automatically).
| `partial_result_set_fetch_size` | Number of intermediate result sets that are buffered in transit for a read query. | False | 1 |
| `ddl_operation_timeout` | Duration in seconds to wait for a DDL operation to complete before timing out | False | 600 seconds |
| `ddl_operation_poll_interval` | Duration in seconds to wait between each polling request for the completion of a DDL operation | False | 5 seconds |

### DDL Statements in v1

> DDL statements in Spanner receive special treatment by Cloud Spanner. Creating and

Reviewer: Are you using the one-line-per-sentence style? I see in other parts of the doc.

Author: Haphazardly. I'll update to that style throughout.
